File: Breadth-First Search/Topological sort/LET_no841_Keys and Rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        queue = collections.deque([0])
        visited = {0}
        
        while queue:
            for _ in range(len(queue)):
                cur = queue.popleft()
                for room in rooms[cur]:
                    if room not in visited:
                        queue.append(room)
                        visited.add(room)
        
        return len(visited) == len(rooms)

File: Breadth-First Search/Topological sort/LET_no841_Keys and Rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        queue<int> q;
        unordered_set<int> visited;
        q.push(0);
        visited.insert(0);

        while (!q.empty()) {
            int n = q.size();
            for (int i = 0; i < n; i++) {
                int cur = q.front();
                q.pop();
                for (int nxt : rooms[cur]) {
                    if (!visited.count(nxt)) {
                        visited.insert(nxt);
                        q.push(nxt);
                    }
                }
            }
        }

        return rooms.size() == visited.size();
    }
};